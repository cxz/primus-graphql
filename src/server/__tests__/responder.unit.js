/* eslint-env jest */
const defaults = require('101/defaults')
const Responder = require('../responder')
const Spark = require('primus/spark')

const defaultOpts = require('../../default-opts')
const primusOpts = require('../../default-primus-opts')

// mock setup
jest.unmock('../responder')
jest.mock('primus/spark')

describe('Responder', () => {
  const ctx = {}

  beforeEach(() => {
    ctx.spark = new Spark()
    ctx.opts = defaults({
      formatError: (err) => err
    }, defaultOpts)
    ctx.primusOpts = defaults({}, primusOpts)
  })

  describe('constructor', () => {
    it('should create a Responder instance', () => {
      const responder = new Responder(ctx.spark, ctx.opts, ctx.primusOpts)
      expect(responder).toBeInstanceOf(Responder)
    })
  })

  describe('methods', () => {
    beforeEach(() => {
      ctx.responder = new Responder(ctx.spark, ctx.opts, ctx.primusOpts)
    })

    describe('send', () => {
      it('should write data to spark', () => {
        const args = {
          id: 'id',
          statusCode: 200,
          payload: {data: { foo: 'foo', bar: 'bar' }}
        }
        ctx.responder.send(args.id, args.statusCode, args.payload)
        expect(ctx.spark.write).toMatchSnapshot()
      })
    })

    describe('sendErrs', () => {
      it('should write error to spark', () => {
        const args = {
          id: 'id',
          statusCode: 400,
          errors: [new Error('boom')]
        }
        ctx.responder.sendErrs(args.id, args.statusCode, args.errors)
        expect(ctx.spark.write).toMatchSnapshot()
      })

      describe('send error', () => {
        beforeEach(() => {
          // stub responder.send
          ctx.responderSend = ctx.responder.send
          ctx.responder.send = () => {
            // unstub after first call
            ctx.responder.send = ctx.responderSend
            throw new Error('send error')
          }
        })
        afterEach(() => {
          // unstub responder.send
          ctx.responder.send = ctx.responderSend
        })

        it('should catch send error and send as error', () => {
          const args = {
            id: 'id',
            statusCode: 400,
            errors: [new Error('boom')]
          }
          ctx.responder.sendErrs(args.id, args.statusCode, args.errors)
          expect(ctx.spark.write).toMatchSnapshot()
        })
      })
    })

    describe('sendEvent', () => {
      it('should write "next" event to spark', () => {
        const args = {
          id: 'id',
          event: 'next',
          data: { foo: 'foo' }
        }
        ctx.responder.sendEvent(args.id, args.event, args.data)
        expect(ctx.spark.write).toMatchSnapshot()
      })

      it('should write "next" event w/ no data to spark', () => {
        const args = {
          id: 'id',
          event: 'next',
          data: { foo: 'foo' }
        }
        ctx.responder.sendEvent(args.id, args.event)
        expect(ctx.spark.write).toMatchSnapshot()
      })

      it('should write "completed" event to spark', () => {
        const args = {
          id: 'id',
          event: 'completed',
          data: { foo: 'foo' }
        }
        ctx.responder.sendEvent(args.id, args.event, args.data)
        expect(ctx.spark.write).toMatchSnapshot()
      })

      it('should write "error" event to spark', () => {
        const args = {
          id: 'id',
          event: 'error',
          data: new Error('boom')
        }
        ctx.responder.sendEvent(args.id, args.event, args.data)
        expect(ctx.spark.write).toMatchSnapshot()
      })

      it('should write "error" event with statusCode to spark', () => {
        const args = {
          id: 'id',
          event: 'error',
          data: defaults(new Error('boom'), { statusCode: 400 })
        }
        ctx.responder.sendEvent(args.id, args.event, args.data)
        expect(ctx.spark.write).toMatchSnapshot()
      })

      it('should write "error" event with multiple errors to spark', () => {
        const args = {
          id: 'id',
          event: 'error',
          data: { errors: [new Error('boom'), new Error('boom2')] }
        }
        ctx.responder.sendEvent(args.id, args.event, args.data)
        expect(ctx.spark.write).toMatchSnapshot()
      })
    })
  })
})
