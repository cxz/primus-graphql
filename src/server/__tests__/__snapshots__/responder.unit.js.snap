// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Responder methods send should write data to spark 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "data": Object {
            "bar": "bar",
            "foo": "foo",
          },
          "id": "id",
          "statusCode": 200,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Responder methods sendErrs send error should catch send error and send as error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "errors": Array [
            [Error: boom],
            [Error: send error],
          ],
          "id": "id",
          "statusCode": 500,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Responder methods sendErrs should write error to spark 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "errors": Array [
            [Error: boom],
          ],
          "id": "id",
          "statusCode": 400,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Responder methods sendEvent should write "completed" event to spark 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "data": Object {
            "foo": "foo",
          },
          "event": "completed",
          "id": "id",
          "statusCode": 200,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Responder methods sendEvent should write "error" event to spark 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "errors": Array [
            [Error: boom],
          ],
          "event": "error",
          "id": "id",
          "statusCode": 500,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Responder methods sendEvent should write "error" event with multiple errors to spark 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "errors": Array [
            [Error: boom],
            [Error: boom2],
          ],
          "event": "error",
          "id": "id",
          "statusCode": 500,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Responder methods sendEvent should write "error" event with statusCode to spark 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "errors": Array [
            [Error: boom],
          ],
          "event": "error",
          "id": "id",
          "statusCode": 400,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Responder methods sendEvent should write "next" event to spark 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "data": Object {
            "foo": "foo",
          },
          "event": "next",
          "id": "id",
          "statusCode": 200,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Responder methods sendEvent should write "next" event w/ no data to spark 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "primus-graphql": Object {
          "event": "next",
          "id": "id",
          "statusCode": 200,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
